Move validations before calling interpret method

qequ opened the pull request:
The method Validations that reports errors on Values was called after the Encoder's interpret function but before encValue, that's why the repeated objects error was reported when exporting to JSON but ignored when exporting to OpenAPI.  Possibly this bug affected ProtobufJSON exporting too.

Moving it before works as expected, signaling the repeated objects error before generating the OpenAPI specification.

OpenAPI tests passed.

#1715

qequ committed:
Move validations of ast before calling interpret function

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

mvdan commented:
This fix seems reasonable, but please add a test :)

To make the DCO bot happy, see https://github.com/cue-lang/cue/blob/master/doc/contribute.md#asserting-a-developer-certificate-of-origin.

myitcv commented:
Welcome to the CUE project, @qequ! 

Please see the [contribution guide](https://github.com/cue-lang/cue/blob/master/doc/contribute.md#good-commit-messages) for advice on commit messages.

qequ committed:
Add tests for validation of OpenApi encoding

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

qequ commented:
@mvdan Added tests :+1:

mvdan commented:
Ah, perhaps I should have clarified - in general, regression tests are added in the form of "testscript" files, which help test end-to-end flows without having to write separate unit tests.

For example, see `cmd/cue/cmd/testdata/script/def_openapi.txt`; you can probably create a similar and smaller sibling of it at `cmd/cue/cmd/testdata/script/export_openapi.txt`.

qequ committed:
cue/cmd/cue/cmd/testdata/script: Add testscript file

Add testscript file export_openapi_err.txt checking eval and export in the case of repeated objects

Fixes #1715

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

qequ commented:
@mvdan Added testscript file

mpvl commented on internal/encoding/encoder_test.go:
```diff
@@ -0,0 +1,65 @@
+	"cuelang.org/go/cue/parser"
+)
+
+func TestEncoderValidation(t *testing.T) {
```
do we still need this test now we have the other test?

mvdan replied:
agreed; that's what I meant originally. this file can be deleted.

mvdan commented on cmd/cue/cmd/testdata/script/export_openapi_err.txt:
```diff
@@ -0,0 +1,22 @@
+cmp stderr expect-stderr
+
+! cue eval repeated.cue --out openapi
+cmp stdout expect-stdout
```
you can just do `! stdout .` to verify that stdout is empty; no need for an empty expect-stdout file.

mvdan reviewed (commented):
I also just realised that export_err.txt exists; maybe you could add your test case as a new sub-package there, to avoid creating a new tiny file.

qequ committed:
internal/enconding: Remove unnecessary test

Fixes #1715

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

qequ committed:
cmd/cue/cmd/testdata/script: Move Repeated objects in openapi validation test

1. Delete file cmd/cue/cmd/testdata/script/export_openapi_err.txt

2. Move tests that were in cmd/cue/cmd/testdata/script/export_openapi_err.txt to cmd/cue/cmd/testdata/script/export_err.txt

Fixes #1715

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

qequ commented:
@mvdan  @mpvl  applied changed to tests

mvdan commented:
Thank you! Could you please squash your commits? Your final commit message is what we will use when importing to Gerrit to then test and merge. See https://github.com/cue-lang/cue/blob/master/doc/contribute.md#good-commit-messages, like Paul said.

qequ committed:
internal/encoding: Fix validations for interpret method in Encoding

Moving Value's Validation method calling before calling interpret method in Encoder's Encode
fixes ignoring errors when exporting to OpenAPI spec such as repeated objects in Values.

Fixes #1715

Signed-off-by: Alvaro Frias Garay <alvaro.frias@eclypsium.com>

qequ committed:
Merge branch 'bugfixRepeatedObjectsError' of github.com:qequ/cue into bugfixRepeatedObjectsError

qequ commented:
@mvdan https://github.com/cue-lang/cue/pull/1716/commits/21576a2b2b6f3621dac80df2f33b47412bbef4a7 squashed commit

mvdan commented:
I still see seven commits. Did you squash and rebase on the latest master?

I can always squash myself and use your commit message from 21576a2b2b6f3621dac80df2f33b47412bbef4a7, if you prefer.

qequ commented:
@mvdan I thought I had. 

> I can always squash myself and use your commit message from https://github.com/cue-lang/cue/commit/21576a2b2b6f3621dac80df2f33b47412bbef4a7, if you prefer.

That would be great! Thank you very much 😁

qequ commented:
@mvdan any updates on this issue?

mvdan commented:
I still intend to merge this; it's just a slightly involved process as we have to import it to Gerrit, and it's the first time I'm doing it - while at the same time automating the process :) I'll likely wrap it up tomorrow.

mvdan commented:
Attempt at https://review.gerrithub.io/c/cue-lang/cue/+/539077, if you want to give it a look. I squashed your commits and reused most of the description you gave in this PR for the commit message.
